File: leetcode/algorithms/distinct_subsequences/distinct_subsequences.py
class Solution(object):
    def numDistinct(self, s, t):
        m = len(s)
        n = len(t)

        counts = [[0 for __ in xrange(n + 1)] for __ in xrange(m + 1)]

        for i in xrange(m):
            counts[i][0] = 1

        for i in xrange(1, m + 1):
            for j in xrange(1, n + 1):
                counts[i][j] += counts[i - 1][j]
                if s[i - 1] == t[j - 1]:
                    counts[i][j] += counts[i - 1][j - 1]

        return counts[m][n]

File: leetcode/algorithms/distinct_subsequences/distinct_subsequences.cc
class Solution {
public:
    int numDistinct(string s, string t) {
        int m = s.size();
        int n = t.size();

        vector<vector<int>> counts(m + 1, vector<int>(n + 1, 0));

        for (int i = 0; i < m; ++i) {
            counts[i][0] = 1;
        }

        for (int i = 1; i <= m; ++i) {
            for (int j = 1; j <= n; ++j) {
                counts[i][j] += counts[i - 1][j];
                if (s[i - 1] == t[j - 1]) {
                    counts[i][j] += counts[i - 1][j - 1];
                }
            }
        }

        return counts[m][n];
    }
};
